source/develop/release-management/features/storage/cd-block-storage.md

```diff
@@ -0,0 +1,228 @@
+                <ovirt-vm:leaseOffset type="int">105906176</ovirt-vm:leaseOffset>
+                <ovirt-vm:leasePath>/dev/88252cf6-381e-48f0-8795-a294a32c7149/leases</ovirt-vm:leasePath>
+                <ovirt-vm:path>/rhev/data-center/mnt/blockSD/88252cf6-381e-48f0-8795-a294a32c7149/images/89f05c7d-b961-4935-993f-514499024515/626a493f-5214-4337-b580-96a1ce702c2a</ovirt-vm:path>
+                <ovirt-vm:volumeID>626a493f-5214-4337-b580-96a1ce702c2a</ovirt-vm:volumeID>
+            </ovirt-vm:volumeChainNode>
+        </ovirt-vm:volumeChain>
+    </ovirt-vm:device>
+```
+
+When CD is not loaded, this element is still present (as CD-ROM is always present), but is empty, e.g.:
+
+```
+    <ovirt-vm:device devtype="disk" name="sdc"/>
+```
+
+### Bugs
+
+- [Bug 1589763: Error changing CD for a running VM when ISO image is on a block domain](https://bugzilla.redhat.com/1589763)
+- [Bug 1868643: Volume not deactivated after ejecting CDROM](https://bugzilla.redhat.com/1868643)
+
+
+### Functionality added with this feature
+
+This feature adds new APIs to vdsm for loading, ejecting and changing CD.
+With these functions, all aforementioned functionality will work properly.
+
+#### Vdsm
+
+- Add VM.insertCD API
+- Add VM.ejectCD API
+- Updated recovery flow to check `in_progress` flag for CDs
+
+
+### Flows
+
+#### VM with empty tray, insert CD
+
+1. engine sends `insertCD` command, specifying ISO by PDIV format
+1. add new drivespec to VM metadata with new attribute `in_progress`
+```
+    <ovirt-vm:device devtype="disk" name="sdc" in_progress="yes">
+```
+1. prepare new drivespec
+  - if failed:
+    1. remove drivespec (log errors)
+    1. fail the flow with original error
+1. loads CD into VM
+  - if failed:
+    1. teardown the volume (log error)
+    1. remove the drive spec (log error)
+    1. fail the flow with the original error
+1. remove `in-progress` attribute from VM metadata
+  - if failed: log the error and succeed
+
+
+#### VM with loaded CD, eject CD
+
+1. engine sends `ejectCD`, specifying ISO path by PDIV format 
+1. vdsm updates VM metadata with new attribute `in_progress`:
+```
+    <ovirt-vm:device devtype="disk" name="sdc" in_progress="yes">
+```
+1. vdsm eject CD from VM (calling libvirt `updateDeviceFlags()`)
+  - if failed (e.g. timeout?):
+    1. remove the `in_progress` flag (log error)
+    1. fail with original error
+1. tear down drivespec
+  - if tearing down volume fails: log error (volume used by another VM)
+1. remove drivespec from VM metadata
+  - if failed: log error and succeed
+
+
+#### VM recovery flow
+
+Try to cleanup storage leftovers from insert and remove operations without changing VM state.
+
+1. metadata in progress, volume is not active
+  - remove metadata from XML
+1. metadata in progress, volume active, no device
+  - deactivate the volume
+  - remove metadata from XML
+1. metadata in progress, volume active, device exists
+  - remove attribute `in_progress` from metadata
```

Comment:
what does the `in_progress` field cover in addition to the state where the medata exists and the device doesn't exist?

Reply:
as mentioned bellow, metadata always exists, as CD is always present. We can skip it, but simplifies the check in which state we are, so unless there are good reason not to have, I'd keep it.

Reply:
metadata exists, <disk> definition exists. 
But the path in source file is what matters.
It does seem redundant to have an extra field. And it requires more metadata updates when used as a persistent store for the flag...We should try to minimize the amount of updates needed, that's an expensive operation.

Reply:
path in the source doesn't express whole state, volume can be active/inactive and path also doesn't say anything about the state when vdsm restarts - if there was some operation in progress which needs cleanup or the CD was already inserted, but we e.g. fail to activate the volume.

Reply:
and how does it help you to have a simple flag in metadata? That's another operations which may fail or vdsm can crash in between and you're have the same problem. Or rather more - now you also have this potentially wron "in_progress" you have to reconcile. (e.g. "you add in_progress to metadata, vdsm crashes right afterwards before you activate")

Yes, you need to find out if the volume is active/inactive when you see a path in disk. But you have to do it anyway. And it's better to check the actual state rather then rely on a flag which suffers from the same problem.

Reply:
if CD is already there (<disk> has a path) but you fail to activate then it's not valid and doesn't really work, right? I don't think it's anything new...right now you can also just use a non-existant path on iso domain (or remove the file in the middle). We don't have that many complaints on that....if it's at VM start then it's dropped (startupPolicy optional). If it's later then it's less nice...but again, we're no handling that today either. You don't have to fix whole vdsm in this effort:)

Reply:
well, so it's fine to remove CD e.g. if the volume is not active no matter if it was successfully inserted before vdsm restart? If only current state matter during recovery, we don't need it.

Reply:
In recovery we always just hope for the best, for normal disks too.
VM runs, it had that path before, presumably working(otherwise it would be dropped due to startupPolicy). If we failk in prepareVolume during recovery we ignore it...because what else can you do, it may in fact keep running fine if your activation error is bogus. Or not. But then again, what would you do...on VM start you fail to start, that's ok, but if it's already running you shouldn't be killing it.

Reply:
The purpose of the in_progress flag is to allow cleanup during recovery. Without this we will have to check every drive spec and device to do the needed cleanup.

If removing it fails after cleaning up stale volumes, we will try agian in the next recovery, there is no hard.

Reply:
I tend to agree with Vojta and Nir here. Using in_progress flag looks like making it more rather than less robust. We don't have such a flag on disk hot plug/unplug though, right?

Reply:
It seems there's more agreement to keep it. I clarified it little bit more in the new version and added example from hot disk plug, which doesn't have it and it can result into active unused volumes.

Reply:
@vjuranek I'd suggest to give it a second thought - as @nirs  pointed out in his last comment, it seems to be possible to achieve that without this flag, right? sure, it may complicate the recovery flow but considering that execute the recovery flow significantly less times than run-vm, I'd say that if we can avoid the extra modification of the metadata on run vm (with cd-rom), it would be the better approach here